### ISSUE TITLE
Scope fixture execution

### ISSUE THREAD

**Issue by pedroreys** [fixie/fixie#96]
Hey Patrick,
This is more a question than an actual issue at the moment.

Let me first describe the problem that I was trying to solve yesterday with @ChrisMissal:

We have two conventions for integration tests, FixturePerMethod and FixturePerClass, with the main distinction being when we reset state between the test runs.

``` csharp
public class FixturePerMethodConvention : IntegrationTestConvention
    {
        public FixturePerMethodConvention()
        {
            Classes
                .ConstructorDoesntHaveArguments();

            ClassExecution
                .CreateInstancePerCase();

             CaseExecution
                .Wrap<IntegrationTestExecutionWrapper>();
        }
}

public class FixturePerClassConvention : IntegrationTestConvention
{
    public FixturePerClassConvention()
    {
        Classes
            .ConstructorHasArguments();

        ClassExecution
            .CreateInstancePerClass()
            .Wrap<IntegrationTestExecutionWrapper>();
    }
}
```

The `IntegrationTestConvention` just filter the `Classes` to the Integration test namespace. It also defines an `ParameterSource` that uses AutoFixture to inject the method arguments.

This all works just fine. The part that we were not able to do is that we wanted to create a NestedContainer scoped to the Fixture, and we tried doing it like this:

``` csharp
public class IntegrationTestExecutionWrapper : CaseBehavior, ClassBehavior
{
    public void Execute(Class context, Action next)
    {
        ResetDataCheckpoint();
        TestContextFixture.Reset();
        next();
    }

    public void Execute(Case context, Action next)
    {
        this.ResetDataCheckpoint();
        TestContextFixture.Reset();
        next();
    }

    private void ResetDataCheckpoint()
    {
        // uses Respawn to reset the db state
    }
}
```

The `TestContextFixture` is a helper class that we can pass to the tests and use to save entities to the DB, send MediatR messages, on my tests, like this:

``` csharp
public void Should_update_customer_name(TestContextFixture fixture, Customer customer)
{
    customer.Name = "Tobias";

    // Saves customer entity to DB
    fixture.SaveAll(customer);

    // send MediatR message
    fixture.Send(new UpdateCustomer { Name = "Mrs. Featherbottom" });

    // Reload entity from DB
    fixture.Reload(ref customer);

    customer.Name.ShouldBe("Mrs. Featherbottom");
}
```

We added that `Reset()` method to the TestContextFixture to create a new scope, i.e to create the nested container.

``` csharp
public void Reset()
{
    this.scope.DisposeCurrentNestedContainer();
    this.scope.CreateNestedContainer(); 
}
```

With that, it was trivial to scope things like the EF DbContext to the NestedContainer. And this worked really well for the `FixturePerClass` tests. But in the `FixturePerMethod` scenario we have a problem that I'm not yet sure the best way to solve.

In our AutoFixture configuration for the Integration Tests, for some entities (usually lookup table entities), we configured AutoFixture to pull an instance from the DB instead of newing one up. For example, we have a "State" table that is just a lookup table for the US States, given that this table doesn't ever change, we configured Respawn to skip it when doing the Checkpoint Reset and have a custom AutoFixture builder that just pulls one random State entity from the DB whenever we need one.

When we added the `Reset` to the TestContextFixture, this became a problem for the `FixturePerMethod` as the method parameter instances were created before the CaseExecution wrapper method ran, essentially making the State entities to be tied to a different DbContext from the one that the TestContextFixture was going to use after the `Reset` call. And this makes EF freak out.

I initially tried a workaround by doing the `Reset` not at the `CaseBehavior` hook but in the `ParameterSource`. That didn't work because Fixie creates the parameter instances for _all_ test methods within a class before executing each test method. Which, again, works fine for our `FixturePerClass` scenario but doesn't work for the `FixturePerMethod` scenario.

Going through this exercise yesterday it felt like I was trying to approach this the wrong way, as I felt I was having to "hack" fixie too much. I believe I'm probably using the wrong extension points to do this. Is there a way that I can reset the scope of the fixture execution, including the test method argument resolution?

**Comment by pedroreys** [comment]
@plioi ,

I was looking into implementing this change but I soon realized that this won't be trivial and will, probably, lead to API changes. So I want to discuss with you before proceeding.

The main reason why this is challenging is due to the fact that `ParameterSource` does the discovery of the case parameters as well as resolve their values at the same time. This means that for cases with method arguments, those arguments get resolved in the `ClassRunner`, when Fixie creates the Case instances.

To make this work, my idea is to change the `ParameterSource` api to allow the separaione of the parameter discovery from the parameter value resolution. I'm thinking something along the lines of:

````csharp
public interface Parameter
{
   public IEnumerable<Func<object[]>> GetParameters(MethodInfo methos);
}
````

That way we can figure out the number of sets of paramenters for a given test method and create all the Case instances, but we can defer the Parameter value resolution to a later stage in the lifecycle, under a Fixture scope.

**Comment by plioi** [comment]
I think returning an IEnumerable<Func<object[]>> alone doesn't help, since we'd *still* have to change *when* the paramter source is *used*.  We may be able to fix where it is used without changing the interface.  I'll take a closer look at the main loop in ClassRunner to see what all would be involved in moving the retrieval of parameters to a more appropriate step of the process.

**Comment by pedroreys** [comment]
:+1:

**Comment by plioi** [comment]
If all of the *sets* of inputs for one test method were resolved right before that test method was called several times in a row, would that be sufficient?  Or, should each individual set of args be resolved right before the single call they belong to?  I bet the latter would be best.

**Comment by pedroreys** [comment]
the latter, yes.

Currently, when you have a `CreateInstancePerCase` lifecycle, and you have multiple sets of parameters, do you create a Fixture scope per each individual set of args?

**Comment by plioi** [comment]
https://github.com/fixie/fixie/blob/master/src/Fixie/Internal/Behaviors/CreateInstancePerCase.cs My "Fixture" class means "test class instance" (which I regret, as I think it causes confusion).  So, yes, when creating an instance per case, you get a whole new Fixture instance for each invocation of the test method.

**Comment by pedroreys** [comment]
What I had in mind was for `Fixture` to be the scope boundary for my tests, according to the lifecycle I have specified, and to be guaranteed that all parameter resolution, both _ctor_ or test method, to happen within this scope boundary.

If I were writing a spec for this here is what I wanted it to look like:

# Fixture Scope

A Fixture scope wraps a Test Class instance. The fixture scope starts right before a Test class is instantiated and is disposed right after the test class is disposed. All arguments, constructor arguments for test class instances and test method arguments for test cases are instantiated within the Fixture scope.

## Fixture Life-cycles
### Instance Per Class

When executing under an Instance per Class lifecycle, all cases defined in a test class are scoped to a single class instance. 

This life-cycle will enable a [Test Class Per Fixture](http://xunitpatterns.com/Testcase%20Class%20per%20Fixture.html) pattern, where the _Setup_ (and usually the _Execution_ as well) is done at the test class constructor and each test case is a different assertions around that same setup and execution.

In a pseudo-code, the scoping would look like:

````csharp
var fixture = startFixtureScope();
var testClassInstance = instantiateTestClass();
foreach(var case in testClass.Cases)
{
    var caseArgs = ResolveCaseArguments(case);
    InvokeCase(testClassInstance, caseArgs);
}
testClassInstance.Dispose();
fixture.Dispose();
````

### Instance Per Case

In an instance per case lifecycle, each case defined in a test class will run under their own test class instance, completely isolated from each other.

This is usually the scenario where each test is completely isolated from the other tests defined in the same test class and usually the setup, execution and assertion parts of the test are defined in the test method body.
In pseudo-code, the scoping would look like:
````csharp
foreach(var case in testClass)
{
      var fixture = startFixtureScope();
      var testClassInstance = instantiateTestClass();
      var caseArgs = ResolveCaseArguments(case);
      InvokeCase(testClassInstance, caseArgs)
      testClassInstance.Dispose();
      fixture.Dispose();
}
````

## Conclusion

That way, regardless of the test class instance lifecycle, one could always create a FixtureBehavior to define the tasks that need to be performed to guarantee the isolated fixture boundary. In the example below, it clears the test database (`checkpoint.Reset()`) and creates a new nested container (`container.CreateNestedContainer`) before proceeding with create the test class instance, resolving case arguments and executing the cases:

````csharp
FixtureExecution.Wrap((fixture, next) =>
{
    checkpoint.Reset();
    container.CreateNestedContainer();
    next();
});
````

**Comment by plioi** [comment]
Thanks for that analysis.  I've been thinking it's time to revisit the test class lifecycle in light of our real-world usages.  If it can be resolved without a breaking change, I'll include it in the 1.x branch.  If not, I have started up a dev branch for breaking changes (so far the only breaking changes have been for people making 3rd party runners, but I anticipate a few convention API changes where needed).

If this falls under the "breaking, 2.x" category of changes, that may make it a longer wait.  How much is this hurting in your current project?

**Comment by pedroreys** [comment]
I have my workarounds. (I currently don't allow Test parameters on integration tests, only ctor arguments).

So, don't feel pressured by my project is what I'm trying to say :wink:

**Comment by webseb** [comment]
It's been a while but did you get to work on that issue since it was reported?

**Comment by plioi** [comment]
@pedroreys With the big change in the customizatoin API in the now-released version 3.0, I believe I finally give enough control over the test project lifecycle, and in such a way where you can handle such things using plain old C# rather than these older complex-and-insufficient 'wrapping behaviors'. I'm closing this issue due to age and the fact that I think the new model gives everyone the control and context they need. Your pain points here definitely influenced the v3 design, so thank you for raising them and for your patience :)

**Comment by pedroreys** [comment]
blast from the past :)
